syntax = "proto3";

option go_package = "gen/lorepb";

package lore;

service LoreService {
  rpc GenerateCharacters (CharactersRequest) returns (CharactersResponse);
  rpc GenerateFactions (FactionsRequest) returns (FactionsResponse);
  rpc GenerateSettings (SettingsRequest) returns (SettingsResponse);
  rpc GenerateEvents (EventsRequest) returns (EventsResponse);
  rpc GenerateRelics (RelicsRequest) returns (RelicsResponse);
  rpc GenerateAll (AllRequest) returns (AllResponse);
  rpc GenerateFullStory (FullStoryRequest) returns (FullStoryResponse);
  rpc GenerateEmbedding (EmbeddingRequest) returns (EmbeddingResponse);
  rpc RerankResults (RerankSearchRequest) returns (RerankSearchResponse);
}

message CharactersRequest {
  string theme = 1;
  int32 count = 2;
  bool regenerate = 3;
}
message FactionsRequest {
  string theme = 1;
  int32 count = 2;
  bool regenerate = 3;
}

message SettingsRequest {
  string theme = 1;
  int32 count = 2;
  bool regenerate = 3;
}

message EventsRequest {
  string theme = 1;
  int32 count = 2;
  bool regenerate = 3;
}

message RelicsRequest {
  string theme = 1;
  int32 count = 2;
  bool regenerate = 3;
}

message LorePiece {
  string name = 1;
  string description = 2;
  map<string, string> details = 3;
  string type = 4;
}

message CharactersResponse {
  repeated LorePiece characters = 1;
}

message FactionsResponse {
  repeated LorePiece factions = 1;
}

message SettingsResponse {
  repeated LorePiece settings = 1;
}

message EventsResponse {
  repeated LorePiece events = 1;
}

message RelicsResponse {
  repeated LorePiece relics = 1;
}

message AllRequest {
  string theme = 1;
  int32 count = 2;
  bool regenerate = 3;
}

message AllResponse {
  repeated LorePiece characters = 1;
  repeated LorePiece factions = 2;
  repeated LorePiece settings = 3;
  repeated LorePiece events = 4;
  repeated LorePiece relics = 5;
}

message SelectedLorePieces {
  LorePiece character = 1;
  LorePiece faction = 2;
  LorePiece setting = 3;
  LorePiece event = 4;
  LorePiece relic = 5;
}

message FullStory {
  string content = 1;
  string theme = 2;
  SelectedLorePieces pieces = 3;
  map<string, string> quest = 4;
}

message FullStoryRequest {
  SelectedLorePieces pieces = 1;
  string theme = 2;
}

message FullStoryResponse {
  FullStory story = 1;
}

message EmbeddingRequest {
  string text = 1;
}

message EmbeddingResponse {
  repeated float embedding = 1;
}

message WorldResult {
  string title = 1;
  string theme = 2;
  string full_story = 3;
  float relevance = 4;
}

message RerankSearchRequest {
  string query = 1;
  repeated WorldResult worlds = 2;
}

message RerankSearchResponse {
  repeated WorldResult reranked_worlds = 1;
}